### PR TITLE
nih/sync endpoint to sync the NIH TCGA/DBGAP whitelist

### PR DESCRIPTION
**Pull request by jmthibault79**
~~Please comment but note that it is completely untested.~~

~~The main commit is "nih/sync endpoint to sync the NIH TCGA/DBGAP whitelist" and the others all involve setting config files.~~

DSDEEPB-2384

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -53,6 +53,7 @@ libraryDependencies ++= {
     "com.typesafe.akka"   %%  "akka-testkit"  % akkaV     % "test",
     "com.typesafe.akka"   %%  "akka-slf4j"    % akkaV,
     ("com.google.api-client" % "google-api-client" % "1.20.0").exclude("com.google.guava", "guava-jdk5"),
+    ("com.google.apis" % "google-api-services-storage" % "v1-rev30-1.20.0").exclude("com.google.guava", "guava-jdk5"),
```
Comment (abaumann) [inline build.sbt]: new thing that popped up for you?  agora had a problem with guava where they had to exclude because the builds were finding duplicate dependency files

File: src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala [hunk 2]
```diff
@@ -88,7 +94,23 @@ class ProfileClientActor(requestContext: RequestContext) extends Actor with Fire
     case GetAndUpdateNIHStatus(userInfo: UserInfo) =>
       getProfile(userInfo, true)
       // do NOT pipe the response back to the parent!
-  }
+
+    case SyncWhitelist =>
+      val (bucket, file) = (FireCloudConfig.Nih.whitelistBucket, FireCloudConfig.Nih.whitelistFile)
+      val whitelist = Source.fromInputStream(HttpGoogleServicesDAO.getBucketObjectAsInputStream(bucket, file)).getLines().toSet
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala]: what about the case when the whitelist does not exist - what happens here?

File: build.sbt (hunk 1 above)
Comment (jmthibault79) [inline build.sbt]: No, this is copy/paste from Rawls.  I have not tested sbt assembly yet - I'll do that now.

File: src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala (hunk 2 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala]: Blows up, probably.  I'll work on a fix.

What have decided as the correct behavior: ignore, ignore X times and then remove access to all, or remove access to all right away?

File: src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala [hunk 3]
```diff
@@ -258,4 +279,41 @@ class ProfileClientActor(requestContext: RequestContext) extends Actor with Fire
...
   }
 
+  // get a mapping of FireCloud user name to NIH User name, for only those Thurloe users with a non-expired NIH link
+  def getCurrentNihUsernameMap: Future[Map[String, String]] = {
+
+    val pipeline = addCredentials(OAuth2BearerToken(HttpGoogleServicesDAO.getAdminUserAccessToken)) ~> sendReceive
+    def getAllUserValuesForKey(key: String): Future[Map[String, String]] = pipeline {
+      Get(s"${UserService.remoteGetQueryURL}key=${key}")
+    } map { response =>
+      if (response.status != OK) throw new Exception(response.toString)   // TODO: better error handling
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala]: Is there something like a standard OrchestrationException that can be thrown here?

Alternatively, something involving ErrorReport and PerRequest.

File: build.sbt (hunk 1 above)
Comment (jmthibault79) [inline build.sbt]: Seems to work fine either way.

File: src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala [hunk 4]
```diff
@@ -258,4 +266,57 @@ class ProfileClientActor(requestContext: RequestContext) extends Actor with Fire
...
+
+    val usernames = getAllUserValuesForKey("linkedNihUsername")
+    val expiretimes = getAllUserValuesForKey("linkExpireTime")
+
+    // compose futures
+    for {
+      usernameKvp <- usernames
+      expirationKvp <- expiretimes
+    } yield {
+      // exclude from results: link expiration is not in Thurloe, value is not a long, value is in the past
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala]: I am not entirely certain that the below works. At the very least all the flatMaps make it hard to follow. What I would probably do is first find all the non-expired fc users:

val validFcUsers = expirationKvp.map{ case (fcUser, expStr) => fcUser -> Try { expStr.toLong }.toOption }.collect {
  case (fcUser, Some(exp: Long)) if DateUtils.now > exp => fcUser
}

Then filter usernameKvp with the results from above:
usernameKvp.filter { case (fcUser, nihUser) => validFcUsers.contains(fcUser) } values

(maybe that all compiles)

**Comment by davidangb** [comment]
fyi on the Travis errors - we have intermittent Travis failues. Just re-run that test once or a few times until it passes and your PR is green (*sigh*)

File: src/main/config/firecloud-orchestration-compose.yaml.ctmpl [hunk 5]
```diff
@@ -15,6 +15,12 @@ app:
     - FIRECLOUD_URL_ROOT={{.Data.env_firecloud_url_root}}
     - GOOGLE_SECRET_JSON={{.Data.env_google_secret_json}}
     - ADMIN_REFRESH_TOKEN={{.Data.env_admin_refresh_token}}
+    - ORCHESTRATION_PEM=/etc/firecloud-account.pem
+    {{$pem_file := printf "secret/dsde/%s/common/firecloud-account.pem" $environment}}
+    {{with vault $pem_file}}
+    - ORCHESTRATION_PEM_CLIENT_ID={{.Data.client_id}}
+    {{end}}
+    - NIH_WHITELIST_BUCKET={{.Data.env_nih_whitelist_bucket}}
```
Comment (davidangb) [inline src/main/config/firecloud-orchestration-compose.yaml.ctmpl]: is this populated in Vault yet?

File: src/main/resources/application.conf [hunk 6]
```diff
@@ -86,12 +86,19 @@ workspace {
...
-  getAll="/thurloe/%s"
   get="/thurloe/%s/%s"
+  getAll="/thurloe/%s"
+  getQuery="/thurloe?"
   delete="/thurloe/%s/%s"
 }
+
+nih {
+  whitelistFile = "users.txt"
+  rawlsGroupName = "dbGapAuthorizedUsers"
```
Comment (davidangb) [inline src/main/resources/application.conf]: this conflicts with #184 but, with the addition of the filename, I like this PR's version better

Comment (rushtong) [inline src/main/resources/application.conf]: I like this version better too. Is it worth it to address in this PR or should we add another story for refactoring the other version?

File: src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala (hunk 3 above)
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala]: nope, nothing standard - use an ErrorReport if you can, or create a new Exception class.

File: src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala (hunk 4 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala]: updated to do this

File: src/main/config/firecloud-orchestration-compose.yaml.ctmpl (hunk 5 above)
Comment (jmthibault79) [inline src/main/config/firecloud-orchestration-compose.yaml.ctmpl]: No.

Comment (jmthibault79) [inline src/main/config/firecloud-orchestration-compose.yaml.ctmpl]: The PEM stuff *is* in Vault.  AFAIK we have not decided on a proper bucket name yet.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala [hunk 7]
```diff
@@ -2,9 +2,12 @@ package org.broadinstitute.dsde.firecloud.dataaccess
 
 import java.io.StringReader
 
-import com.google.api.client.googleapis.auth.oauth2.{GoogleAuthorizationCodeFlow, GoogleClientSecrets, GoogleTokenResponse}
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: uh oh now you've seen how much I bastardized rawls' version of HttpGoogleServicesDAO!

**Comment by davidangb** [comment]
code structure looks good overall. Let me know when it's ready for testing. Also, looks like this needs a rebase against develop.

**Comment by davidangb** [comment]
Can you factor out any of the methods so they are testable, either on their own or with mock services for Thurloe, etc? Specifically I'm thinking about the code that reads the whitelist, queries Thurloe, and generates the list of valid users.

**Comment by jmthibault79** [comment]
rebased.  will refactor after my manual testing works.

**Comment by jmthibault79** [comment]
@abaumann brought up a good point: this needs to move outside `/api` and/or somewhere that doesn't require calling it "as a user".  I am not familiar with how to do this.

File: src/main/config/firecloud-orchestration-compose.yaml.ctmpl [hunk 8]
```diff
@@ -15,6 +15,13 @@ app:
     - FIRECLOUD_URL_ROOT={{.Data.env_firecloud_url_root}}
     - GOOGLE_SECRET_JSON={{.Data.env_google_secret_json}}
     - ADMIN_REFRESH_TOKEN={{.Data.env_admin_refresh_token}}
+    - REFRESH_TOKEN_SECRET_JSON={{.Data.env_refresh_token_secret_json}}
```
Comment (davidangb) [inline src/main/config/firecloud-orchestration-compose.yaml.ctmpl]: in Vault's dev env, env_refresh_token_secret_json and env_google_secret_json are equivalent. If they are intended to be the same, let's just use one? If they are intended to be different, Vault dev needs updating.

Comment (jmthibault79) [inline src/main/config/firecloud-orchestration-compose.yaml.ctmpl]: It's not clear to me how all the different credentials interact, so I would like to keep them separate for today, and revisit post-release.

**Comment by jmthibault79** [comment]
Please re-review.

Some basic unit tests, and verified to work manually.

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/NIHWhitelistSpec.scala [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  val currentUsernames1 = Map("fcSubjectId2" -> "nihUsername2")
+
+  val expiretimes2 = Future { Map("fcSubjectId1" -> DateUtils.nowMinus24Hours.toString) }
+  val expiretimes3 = Future { Map("fcSubjectId1" -> DateUtils.nowMinus24Hours.toString, "fcSubjectId2" -> "not a number") }
+
+  "NihWhitelistSpec" - {
+    "when handling expired and unexpired users" - {
+      "should only include unexpired users" in {
+        assertResult(currentUsernames1) {
+          Await.result(NIHWhitelistUtils.filterForCurrentUsers(usernames, expiretimes1), 1 second)
```
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/NIHWhitelistSpec.scala]: These give compile warnings when running tests, pretty simple fix:

```
[warn] /Users/davidan/work/src/firecloud-orchestration/src/test/scala/org/broadinstitute/dsde/firecloud/service/NIHWhitelistSpec.scala:24: postfix operator second should be enabled
[warn] by making the implicit value scala.language.postfixOps visible.
[warn] This can be achieved by adding the import clause 'import scala.language.postfixOps'
[warn] or by setting the compiler option -language:postfixOps.
[warn] See the Scala docs for value scala.language.postfixOps for a discussion
[warn] why the feature should be explicitly enabled.
[warn]           Await.result(NIHWhitelistUtils.filterForCurrentUsers(usernames, expiretimes1), 1 second)
[warn]                                                                                            ^
[warn] /Users/davidan/work/src/firecloud-orchestration/src/test/scala/org/broadinstitute/dsde/firecloud/service/NIHWhitelistSpec.scala:32: postfix operator second should be enabled
[warn] by making the implicit value scala.language.postfixOps visible.
[warn]           Await.result(NIHWhitelistUtils.filterForCurrentUsers(usernames, expiretimes2), 1 second)
[warn]                                                                                            ^
[warn] /Users/davidan/work/src/firecloud-orchestration/src/test/scala/org/broadinstitute/dsde/firecloud/service/NIHWhitelistSpec.scala:40: postfix operator second should be enabled
[warn] by making the implicit value scala.language.postfixOps visible.
[warn]           Await.result(NIHWhitelistUtils.filterForCurrentUsers(usernames, expiretimes3), 1 second)
[warn]                                                                                            ^
[warn] three warnings found
```

File: src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala [hunk 10]
```diff
@@ -273,4 +283,64 @@ class ProfileClientActor(requestContext: RequestContext) extends Actor with Fire
       postResponse.entity.asString.contains(Conflict.reason)
   }
 
+  def syncWhitelist: Unit = {
+    val (bucket, file) = (FireCloudConfig.Nih.whitelistBucket, FireCloudConfig.Nih.whitelistFile)
+    val whitelist = Source.fromInputStream(HttpGoogleServicesDAO.getBucketObjectAsInputStream(bucket, file)).getLines().toSet
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala]: here, or maybe in HttpGoogleServicesDAO, trap and catch any exceptions in reading the whitelist from the bucket? My concern here is that if we let the raw exception bubble up, we might be exposing sensitive information (bucket name, or even worse some of the secret credentials) in a stack trace visible by the end user.

File: src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala [hunk 11]
```diff
@@ -273,4 +283,64 @@ class ProfileClientActor(requestContext: RequestContext) extends Actor with Fire
...
+    val url = FireCloudConfig.Rawls.overwriteGroupMembershipUrlFromGroupName(FireCloudConfig.Nih.rawlsGroupName)
+
+    memberList flatMap { members => pipeline { Put(url, members) } }
+  }
+
 }
+
+object NIHWhitelistUtils {
+  def getAllUserValuesForKey(key: String)(implicit arf: ActorRefFactory, ec: ExecutionContext): Future[Map[String, String]] = {
+    val pipeline = addCredentials(OAuth2BearerToken(HttpGoogleServicesDAO.getAdminUserAccessToken)) ~> sendReceive
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala]: future refactoring ... this code:
```
addCredentials(OAuth2BearerToken(HttpGoogleServicesDAO.getAdminUserAccessToken))
```
could/should be encapsulated into a custom RequestTransformer in FireCloudRequestBuilding, something like "addAdminCredentials"

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala [hunk 12]
```diff
@@ -100,4 +110,29 @@ object HttpGoogleServicesDAO {
 
   def randomString(length: Int) = scala.util.Random.alphanumeric.take(length).mkString
 
+  def getAdminUserAccessToken = {
+    val googleCredential = new GoogleCredential.Builder().setTransport(httpTransport)
+      .setJsonFactory(jsonFactory)
+      .setClientSecrets(refreshTokenClientSecrets)
+      .build().setFromTokenResponse(new TokenResponse().setRefreshToken(adminUserRefreshToken))
+
+    googleCredential.refreshToken()
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: is this line necessary?

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/NIHWhitelistSpec.scala [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        }
+      }
+    }
+
+    "when handling users with no expiration times" - {
```
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/NIHWhitelistSpec.scala]: nitpick - this tests users who are expired, not users with no expiration times

Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/NIHWhitelistSpec.scala]: The purpose of this test is to show what happens when a user is in `usernames` but not in `expiretimes`.  It also shows that an expired user is not returned, but this is true for all of the tests.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala (hunk 12 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: Yes.  This is the line that actually generates the new access token.

**Comment by jmthibault79** [comment]
Addressed review comments

**Comment by davidangb** [comment]
okay, @jmthibault79 I'm giving this a big ol' :+1: . Please squash and rebase if needed. Feel free to merge on your own.
